해시태그 팔로우 구현, 검색 API 개선

vectorch9 opened the pull request:
## 해시태그 팔로우
### 해시태그 팔로우 구현
- 팔로우, 언팔로우 API 구현
- 피드에 팔로우한 해시태그의 게시물을 표시하는 것은 추가 작업필요
> 피드에 해시태그의 모든 게시물을 노출하는 것이 아닌 "인기 게시물"만 노출하는 것 같음
## 검색 API 개선
### 기존의 유저만 검색하던 API 삭제
### 해시태그 연관 검색어 목록 페이징 조회 API 제거
### 신규 검색 API
- 검색 조회수(count)를 기준으로 상위 50개의 결과 반환
- 이때, 검색어와 정확히 일치하는 해시태그와 유저(username기준)는 결과에 반드시 포함됨
- 회원가입 시 `search` 테이블에 `count`가 0 인 상태로 회원 정보 추가
- 회원탈퇴는 아직 구현하지 않아서 이후 추가 구현할 예정 #144 
- 해시태그의 경우 `search` 테이블에 등록하는 로직이 아직 없어 쿼리를 통해 DB에 직접 `insert`하여 테스트 했음
-> 해당 로직 구현 후 한번더 테스트 부탁드립니다 @seonpilKim 
- 검색어 앞에 `#`이 붙은 경우엔 해시태그만 검색됨
### 검색 조회수 증가 API
- 해시태그, 유저 각각의 API로 구현했음
- 검색 후 클릭시, 해당 프로필로 이동하기 전 검색 조회수 증가 API를 호출하면 됨
## 게시물-해시태그 등록 로직 업데이트
- 새로운 해시태그를 등록하는 경우, SearchHashtag 엔티티 생성
- 해시태그를 삭제해야하는 경우, SearchHashtag 엔티티 제거(batch delete)

## 해결 이슈
- Resolve: #136 
- Resolve: #143

vectorch9 committed:
Feat: 해시태그 팔로우 구현

vectorch9 committed:
Merge remote-tracking branch 'origin/develop' into Feature/HashtagFollow

vectorch9 committed:
Fix: 기존의 유저 검색 API 제거

vectorch9 committed:
Feat: 유저, 해시태그 검색 API 추가

seonpilKim commented on src/main/java/cloneproject/Instagram/dto/search/SearchDTO.java:
```diff
@@ -0,0 +1,22 @@
+@Getter
+@Setter
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@AllArgsConstructor(access = AccessLevel.PROTECTED)
```
의도적으로 @AllArgsConstructor를 protected로 지정하신 건가요?

vectorch9 replied:
SearchDTO를 단독으로 사용할 수 없도록 하려고 protected로 지정했습니다. abstract 키워드도 추가하는게 좋을까요 ?

seonpilKim replied:
@bluetifulc 네네 단독으로 인스턴스를 생성하는게 아니면 추상클래스로 만드는게 좋을 것 같아요

seonpilKim commented on src/main/java/cloneproject/Instagram/repository/search/SearchRepositoryQuerydslImpl.java:
```diff
@@ -0,0 +1,263 @@
+            ))
+        );
+
+        boolean matchingMemberExists = queryFactory
+            .from(searchMember)
+            .innerJoin(searchMember.member, member).fetchJoin()
+            .where(searchMember.id.in(searchIds).and(searchMember.member.username.eq(text)))
+            .fetchFirst() != null;
+        boolean matchingHashtagExists = queryFactory
+            .from(searchHashtag)
+            .innerJoin(searchHashtag.hashtag, hashtag).fetchJoin()
+            .where(searchHashtag.id.in(searchIds).and(searchHashtag.hashtag.name.eq(text)))
+            .fetchFirst() != null;
```
각각 memberMap, hashtagMap에서도 찾을 수 있지 않을까요?

vectorch9 replied:
쿼리를 실행하는 것보단 List의 `contains`와 같은 함수를 사용하는게 나을까요?

seonpilKim replied:
@bluetifulc db랑 통신하는 것 보다 메모리에서 연산하는게 훨씬 빠른 거로 알고 있어요.

seonpilKim commented on src/main/java/cloneproject/Instagram/repository/search/SearchRepositoryQuerydslImpl.java:
```diff
@@ -0,0 +1,263 @@
+        boolean matchingHashtagExists = queryFactory
+            .from(searchHashtag)
+            .innerJoin(searchHashtag.hashtag, hashtag).fetchJoin()
+            .where(searchHashtag.id.in(searchIds).and(searchHashtag.hashtag.name.eq(text)))
```
hashtagMap에서 찾을 수도 있지 않을까요?

seonpilKim reviewed (commented):
고생하셨습니다. 제가 작업할 부분은 이따 저녁에 수정하겠습니다.
+ 이슈들 PR에 연결해주세요~

seonpilKim commented on src/main/java/cloneproject/Instagram/controller/SearchController.java:
```diff
@@ -0,0 +1,62 @@
+        return new ResponseEntity<>(result, HttpStatus.valueOf(result.getStatus()));
+    }
+
+    @ApiOperation(value = "유저 검색 조회수 증가")
+    @ApiImplicitParam(name = "username", value = "조회수 증가시킬 username", required = true, example = "dlwlrma")
+    @PostMapping(value = "/topsearch/member/upcount")
+    public ResponseEntity<ResultResponse> increaseSearchMemberCount(@RequestParam String username) {
+        searchService.increaseSearchMemberCount(username);
+
+        ResultResponse result = ResultResponse.of(ResultCode.INCREASE_SEARCH_COUNT_SUCCESS, null);
+        return new ResponseEntity<>(result, HttpStatus.valueOf(result.getStatus()));
+    }
+
+    @ApiOperation(value = "해시태그 검색 조회수 증가")
+    @ApiImplicitParam(name = "name", value = "조회수 증가시킬 hashtag", required = true, example = "만두")
+    @PostMapping(value = "/topsearch/hashtag/upcount")
+    public ResponseEntity<ResultResponse> increaseSearchHashtagCount(@RequestParam String name) {
+        searchService.increaseSearchHashtagCount(name);
+
+        ResultResponse result = ResultResponse.of(ResultCode.INCREASE_SEARCH_COUNT_SUCCESS, null);
+        return new ResponseEntity<>(result, HttpStatus.valueOf(result.getStatus()));
+    }
```
이 부분은 요청 데이터에 type{member, hashtag, ...}을 추가해서 하나의 API로 요청받고, 서비스단에서 type에 따라 분기문으로 처리하면 어떨까요?
인스타는 `https://i.instagram.com/api/v1/fbsearch/register_recent_search_click/`으로 아래와 같이 요청하더라구요.
```
entity_id: 123124124,
entity_type: user
```

seonpilKim commented on src/main/java/cloneproject/Instagram/controller/SearchController.java:
```diff
@@ -0,0 +1,62 @@
+        return new ResponseEntity<>(result, HttpStatus.valueOf(result.getStatus()));
+    }
+
+    @ApiOperation(value = "유저 검색 조회수 증가")
+    @ApiImplicitParam(name = "username", value = "조회수 증가시킬 username", required = true, example = "dlwlrma")
+    @PostMapping(value = "/topsearch/member/upcount")
+    public ResponseEntity<ResultResponse> increaseSearchMemberCount(@RequestParam String username) {
+        searchService.increaseSearchMemberCount(username);
+
+        ResultResponse result = ResultResponse.of(ResultCode.INCREASE_SEARCH_COUNT_SUCCESS, null);
+        return new ResponseEntity<>(result, HttpStatus.valueOf(result.getStatus()));
+    }
+
+    @ApiOperation(value = "해시태그 검색 조회수 증가")
+    @ApiImplicitParam(name = "name", value = "조회수 증가시킬 hashtag", required = true, example = "만두")
+    @PostMapping(value = "/topsearch/hashtag/upcount")
+    public ResponseEntity<ResultResponse> increaseSearchHashtagCount(@RequestParam String name) {
+        searchService.increaseSearchHashtagCount(name);
+
+        ResultResponse result = ResultResponse.of(ResultCode.INCREASE_SEARCH_COUNT_SUCCESS, null);
+        return new ResponseEntity<>(result, HttpStatus.valueOf(result.getStatus()));
+    }
```
이 부분은 요청 데이터에 type{member, hashtag, ...}을 추가해서 하나의 API로 요청받고, 서비스단에서 type에 따라 분기문으로 처리하면 어떨까요?
인스타는 `https://i.instagram.com/api/v1/fbsearch/register_recent_search_click/`으로 아래와 같이 요청하더라구요.
```
entity_id: 123124124,
entity_type: user
```

vectorch9 replied:
넵 위의 memberMap, hashMap과 함께 수정해서 커밋 올리겠습니다

seonpilKim committed:
Feat: 해시태그 연관 검색어 목록 페이징 조회 API 제거

Resolve: #143

seonpilKim committed:
Feat: 게시물/댓글 업로드 or 삭제 API 로직 추가

- 게시물/댓글 업로드 API: 해시태그 첫 생성 시, SearchHashtag 엔티티 생성
- 게시물/댓글 삭제 API: 해시태그를 삭제하는 경우, SearchHashtag 엔티티 제거

Resolve: #143

vectorch9 committed:
Fix: 일치하는 검색어 처리 로직 수정

- 검색어와 완전히 일치하는 해시태그, 유저네임의 처리 수정
    - 쿼리를 날리는 방식에서 List의 contains함수를 이용하도록 수정
- 검색 조회수 증가의 유저, 해시태그를 하나의 API로 합침

vectorch9 committed:
Fix: 테스트 후 복구시켜놓지 않은 코드 수정

vectorch9 commented:
### 일치하는 검색어 처리 로직 수정
- 검색어와 일치하는 유저네임, 해시태그 처리 로직 수정
- 쿼리를 통해 점검하는 방식 -> List의 `contains` 함수를 이용하는 방식
### 검색 조회수 증가 API 수정
- 유저, 해시태그 각각의 API를 하나의 API로 합침

seonpilKim commented on src/main/java/cloneproject/Instagram/controller/SearchController.java:
```diff
@@ -39,21 +40,14 @@
+        @ApiImplicitParam(name = "entityName", value = "조회수 증가시킬 식별 name", required = true, example = "dlwlrma"),
+        @ApiImplicitParam(name = "entityType", value = "조회수 증가시킬 type", required = true, example = "MEMBER")
+    })
     @PostMapping(value = "/topsearch/member/upcount")
```
/topsearch/upcount로 변경하는게 좋을 것 같아요.

vectorch9 replied:
수정했습니다 !

vectorch9 committed:
Fix: 검색 조회수 증가 URI 변경
